### DISCUSSION TITLE
How to implement undo/redo toolbar buttons?

### DISCUSSION THREAD

**Discussion by AM-thedev** [ianstormtaylor/slate#5303] [Q&A]
How would I implement undo/redo buttons on the toolbar in, for example, the Rich Text Editor demo?
I've searched everywhere but there doesn't seem to be any documentation on how to implement slate-history as a button.

**Comment by klingat** [comment]
I was looking for this too. 

This is what worked for me:
1. Install the plugin: `yarn add slate-history`
2. Import `withHistory`:
```js
 import { withHistory } from 'slate-history';
```
 
3. Initialize editor with the `withHistory` provider:
```js
  const [editor] = useState(() => withReact(withHistory(createEditor())));
  ```
  
4. Then add buttons for undo/redo:
```js
  const initialValue = [
    {
      type: 'paragraph',
      children: [{ text: 'A line of text in a paragraph.' }],
    },
  ];

  return (
    <>
      <button onClick={editor.undo}>Undo</button>
      <button onClick={editor.redo}>Redo</button>
      <Slate editor={editor} value={initialValue}>
        <Editable />
      </Slate>
    </>
  );
  ```

That should work!

**Reply by maylortaylor** [reply]
This should be in the Docs somewhere
